src/py/flwr/server/state/sqlite_state.py

```diff
@@ -0,0 +1,582 @@
+# Copyright 2023 Adap GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""SQLite based implemenation of server state."""
+import re
+import sqlite3
+from datetime import datetime, timedelta, timezone
+from logging import ERROR
+from typing import Any, Dict, List, Optional, Set, Tuple, Union, cast
+from uuid import UUID, uuid4
+
+from flwr.common.logger import log
+from flwr.proto.node_pb2 import Node
+from flwr.proto.task_pb2 import Task, TaskIns, TaskRes
+from flwr.proto.transport_pb2 import ClientMessage, ServerMessage
+from flwr.server.utils.validator import validate_task_ins_or_res
+
+from .state import State
+
+SQL_CREATE_TABLE_NODE = """
+CREATE TABLE IF NOT EXISTS node(
+    id INTEGER UNIQUE
+);
+"""
+
+SQL_CREATE_TABLE_TASK_INS = """
+CREATE TABLE IF NOT EXISTS task_ins(
+    task_id                 TEXT UNIQUE,
+    group_id                TEXT,
+    workload_id             TEXT,
+    producer_anonymous      BOOLEAN,
+    producer_node_id        INTEGER,
+    consumer_anonymous      BOOLEAN,
+    consumer_node_id        INTEGER,
+    created_at              TEXT,
+    delivered_at            TEXT,
+    ttl                     TEXT,
+    ancestry                TEXT,
+    legacy_server_message   BLOB,
+    legacy_client_message   BLOB
+    -- FOREIGN KEY(task_id) REFERENCES node(id)
```

Comment:
This doesn't seem right, `task_id` is not a foreign key, but `producer_node_id` and `consumer_node_id` are (in cases where the node is not anonymous).

However, given that we can register and unregister nodes, should we perhaps just not add any foreign key constraint here?

Reply:
The `--` basically means its a comment. I think I had it in there to just test it, but than also thought right now we can't have an FK as we simply remove the node id from the table. I will remove the SQL comments.